File: DisplayingGraphicsDemo/app/src/main/java/gpj/com/displayinggraphics/MyGLSurfaceView.java
package gpj.com.displayinggraphics;

import android.content.Context;
import android.opengl.GLSurfaceView;
import android.view.MotionEvent;

class MyGLSurfaceView extends GLSurfaceView {

    private final MyGLRenderer mRenderer;

    public MyGLSurfaceView(Context context){
        super(context);

        // 创建一个OpenGL ES 2.0 的context
        setEGLContextClientVersion(2);

        mRenderer = new MyGLRenderer();

        // 设置渲染器（Renderer）以在GLSurfaceView上绘制
        setRenderer(mRenderer);

        // 仅在绘图数据发生更改时才渲染视图
        setRenderMode(GLSurfaceView.RENDERMODE_WHEN_DIRTY);
    }

    private final float TOUCH_SCALE_FACTOR = 180.0f / 320;
    private float mPreviousX;
    private float mPreviousY;

    @Override
    public boolean onTouchEvent(MotionEvent e) {

        // MotionEvent报告触摸屏和其他输入控件的输入详细信息。
        // 在这种情况下，这里只对触摸位置发生变化的事件感兴趣。
        float x = e.getX();
        float y = e.getY();

        switch (e.getAction()) {
            case MotionEvent.ACTION_MOVE:

                float dx = x - mPreviousX;
                float dy = y - mPreviousY;

                // reverse direction of rotation above the mid-line
                if (y > getHeight() / 2) {
                    dx = dx * -1 ;
                }

                // reverse direction of rotation to left of the mid-line
                if (x < getWidth() / 2) {
                    dy = dy * -1 ;
                }

                mRenderer.setAngle(
                        mRenderer.getAngle() +
                                ((dx + dy) * TOUCH_SCALE_FACTOR));
                requestRender();
        }

        mPreviousX = x;
        mPreviousY = y;
        return true;
    }
}

File: Transformations/app/src/main/java/gpj/com/transformations/MyGLSurfaceView.kt
package gpj.com.transformations

import android.content.Context
import android.opengl.GLSurfaceView

internal class MyGLSurfaceView(context: Context) : GLSurfaceView(context) {

    private val mRenderer: MyGLRenderer

    init {

        // 创建一个OpenGL ES 2.0 的context
        setEGLContextClientVersion(3)

        mRenderer = MyGLRenderer(this)

        // 设置渲染器（Renderer）以在GLSurfaceView上绘制
        setRenderer(mRenderer)

        // 仅在绘图数据发生更改时才渲染视图
        renderMode = GLSurfaceView.RENDERMODE_WHEN_DIRTY
    }


}
